packages/client/src/clients/in-app-messages/message-client.ts

```diff
@@ -133,11 +137,129 @@ export class InAppMessagesClient {
     };
   }
 
+  // ----------------------------------------------
+  // Message engagement
+  // ----------------------------------------------
+  async markAsSeen(itemOrItems: InAppMessage | InAppMessage[]) {
+    const itemIds = this.getItemIds(itemOrItems);
+
+    this.channelClient.setMessageAttrs(itemIds, {
+      seen_at: new Date().toISOString(),
+    });
+
+    return this.makeStatusUpdate(itemOrItems, "seen");
+  }
+
+  async markAsUnseen(itemOrItems: InAppMessage | InAppMessage[]) {
+    const itemIds = this.getItemIds(itemOrItems);
+
+    this.channelClient.setMessageAttrs(itemIds, {
+      seen_at: null,
+    });
+
+    return this.makeStatusUpdate(itemOrItems, "unseen");
+  }
+
+  async markAsRead(itemOrItems: InAppMessage | InAppMessage[]) {
+    const itemIds = this.getItemIds(itemOrItems);
+
+    this.channelClient.setMessageAttrs(itemIds, {
+      read_at: new Date().toISOString(),
+    });
+
+    return this.makeStatusUpdate(itemOrItems, "read");
+  }
+
+  async markAsUnread(itemOrItems: InAppMessage | InAppMessage[]) {
+    const itemIds = this.getItemIds(itemOrItems);
+
+    this.channelClient.setMessageAttrs(itemIds, {
+      read_at: null,
+    });
+
+    return this.makeStatusUpdate(itemOrItems, "unread");
+  }
+
+  async markAsInteracted(
+    itemOrItems: InAppMessage | InAppMessage[],
+    metadata?: Record<string, string>,
+  ) {
+    const now = new Date().toISOString();
+    const itemIds = this.getItemIds(itemOrItems);
+
+    this.channelClient.setMessageAttrs(itemIds, {
+      read_at: now,
+      interacted_at: now,
+    });
+
+    return this.makeStatusUpdate(itemOrItems, "interacted", metadata);
+  }
+
+  async markAsArchived(itemOrItems: InAppMessage | InAppMessage[]) {
+    const itemIds = this.getItemIds(itemOrItems);
+
+    const queryInfo = this.channelClient.store.state.queries[this.queryKey];
+
+    // Optimistically update the message status.
+    this.channelClient.setMessageAttrs(itemIds, {
+      archived_at: new Date().toISOString(),
+    });
+
+    // If set to exclude, also remove the message id from the list of messages for the given query
```

Comment:
Hm I'm not sure about this - whether it's correct to change the query result info after the fact. And even if it is desirable, i think it'd be hard to handle that behavior correctly..? For example, if you had set a filter param of engagement status = unseen, then calling `markAsSeen()` implies you would have to make a similar operation here? Not only that you can also have multiple engagement status filters at the same time, which would makes things quite complex. And what happens if you reverse it by calling `markAsUnSeen()` for the same message - do we then put the message id back in there? 🤔

Reply:
I'll double check what the feed does for the reverse operations (like mark unarchived), but we do need this for optimistic updates, otherwise we'd have to wait for another fetch (or ws message).

Reply:
Going to comment this out for now and leave it as a todo. Will bring it up with the team